salt/modules/keyboard.py

```diff
@@ -63,21 +63,28 @@ def set_sys(layout):
 
         salt '*' keyboard.set_sys dvorak
     """
+    result = None
     if salt.utils.path.which("localectl"):
-        __salt__["cmd.run"]("localectl set-keymap {}".format(layout))
+        result = __salt__["cmd.retcode"]("localectl set-keymap {}".format(layout))
+    #TODO: Make behaviour dependent on outcome of command below
+    # Right now the "result = 0" just mimics the old behaviour, to stop things
+    # breaking, but it's not the right thing to do as it ignores the retcode.
     elif "RedHat" in __grains__["os_family"]:
         __salt__["file.sed"](
             "/etc/sysconfig/keyboard", "^LAYOUT=.*", "LAYOUT={}".format(layout)
         )
+        result = 0
     elif "Debian" in __grains__["os_family"]:
         __salt__["file.sed"](
             "/etc/default/keyboard", "^XKBLAYOUT=.*", "XKBLAYOUT={}".format(layout)
         )
+        result = 0
     elif "Gentoo" in __grains__["os_family"]:
         __salt__["file.sed"](
             "/etc/conf.d/keymaps", "^keymap=.*", "keymap={}".format(layout)
         )
-    return layout
+        result = 0
+    return result
```

Comment:
This is changing behavior. If the commands where a success we should still return `layout` which is what the previous behavior was.

Reply:
Ah sorry, yeah, it does change what's returned from that function and might break things for someone who's using the module function directly. There's nothing I could find that does in the salt codebase, so I'm not sure if that's likely, but someone might have built their own tools that would be affected.

What I mean is that from the user's perspective, it doesn't change what happens when the user runs the command to set the keyboard layout. If I kept it returning layout, the calling function would now fail, because the calling function now expects this function to return a return code. It's confusing to write out but I ended up with this:

Previously the state function called the module function and gave it a layout. if the module function finished running, it would still have something saved as 'layout', so it would give that back to the state function. the state function would check it got anything at all back from the module function, and if it did, it would report success. It doesn't care if the layout is actually set, or if the layout variable contains anything valid; it just has to contain *something*. So in practice, this just becomes a check for if the function runs all the way without erroring.

So although it returns 'layout', it doesn't do anything with it and you could use any variable at all that the calling function provided. (Or check for a return code directly.)

I put these here so that if an end user on gentoo/redhat/debian ran the set-keyboard commands, they wouldn't see any difference in behaviour to before, instead of breaking it for them. But it would be better to actually _fix_ it for them. :) I think the correct behaviour is for all of them to return 0 iff the layout is set.

Reply:
I agree there is not a alot of value in returning `layout`, but we should still properly deprecate that the return of this function is going to change.

And instead of returning 0, would it make more sense to return True vs False instead, which is inline with a lot of other salt functions.

Let me get some more memebers of my team's opinions on the deprecating the return and what we should return.

@whytewolf @garethgreenaway thoughts here?

Reply:
Thank you both for the feedback! Okay, I'll try to fix that.

Reply:
Thanks again for your help so far! :) I've adjusted it to return True instead of 0, and updated the unit tests to match. I haven't yet been able to test the behaviour when localectl is unavailable (I currently expect I'll be able to do that for redhat and debian, but it may be harder for gentoo). I'm also not sure what I need to do in order to follow the deprecation path with the return. I found https://docs.saltproject.io/en/latest/topics/development/deprecations.html ; does this mean I should make a separate patch to merge first, that will just contain the warning?

Reply:
Normally we add a warning in the function that will spit out a deprecation message. But I think this is going to be too noisy since we are just letting users now that the return of this function will be changing in an upcoming release.

How about just adding an additional changelog file with the file extension `.deprecated` and include a message that
the `set_sys` and `get_x` functions in the keyboard module will change in the Argon release to return True instead of the value of `layout`.

Then we should be good to go forward with PR.

Reply:
@zzaimeche just making sure you saw my feedback ^

Reply:
@Ch3LL Thanks! I can add that file, then. :)